cli/cli/cmds_volume.go

```diff
@@ -339,7 +344,11 @@ func (c *CLI) initVolumeCmds() {
 			// volumes that are attached or can be attached should be valid
 			// candidates for a mount operation
 			result, err := c.lsVolumes(
-				args, volumeStatusAttached, volumeStatusAvailable)
+				args,
+				requestAttachments(
+					apitypes.VolumeAttachmentsMine,
+					apitypes.VolumeAttachmentsAttached,
+					apitypes.VolumeAttachmentsUnattached))
```

Comment:
It does make sense. However, the masks `3` and `11` are **intended** to return different results.

A mask of `3` should return **all** volumes, but only bother fetching/returning the attachment information for the instance specified in the instance ID header.

A mask of `11` should return **only** the volumes attached to the instance specified in the instance ID header. I worded the description poorly as technically a driver doesn't have to retrieve info for all volumes if the backend has an API that allows it to return attachment info for a particular instance. However, I was thinking along the lines of how you likely have to implement it -- get all volumes and their attachment info, filter out only the volumes that are attached to a provided instance.

Essentially the **intended** difference between `3` and `11` is that in `3` all volumes are returned, but only the ones attached to the provided instance ID have attachment info. For mask `11` only the volumes attached to the provided instance ID have attachment info and are the only ones returned.

Let's look at the code to which you refer:

```go
// if only the requesting instance's attachments are requested then
	// filter the volume's attachments list
	if attachments.Mine() {
		atts := []*types.VolumeAttachment{}
		for _, a := range vol.Attachments {
			alf := log.Fields{
				"attDeviceName": a.DeviceName,
				"attDountPoint": a.MountPoint,
				"attVolumeID":   a.VolumeID,
			}
			if strings.EqualFold(iid.ID, a.InstanceID.ID) {
				atts = append(atts, a)
				ctx.WithFields(lf).WithFields(alf).Debug(
					"including volume attachment")
			} else {
				ctx.WithFields(lf).WithFields(alf).Debug(
					"omitting volume attachment")
			}
		}
		vol.Attachments = atts
		ctx.WithFields(lf).Debug("included volume attached to instance")
	}
```

The interior of the `Mine()` scope doesn't omit the **volume** -- only its attachment(s). If the `2` bit is set, only attachments attached to the provided instance are retained in the attachments list. That's separate and unrelated to whether a volume is returned based on the presence of bits `8` or `16`:

```go
	// if the volume has no attachments and the mask indicates that
	// only attached volumes should be returned then omit this volume
	if len(vol.Attachments) == 0 &&
		attachments.Attached() &&
		!attachments.Unattached() {
		ctx.WithFields(lf).Debug("omitting unattached volume")
		return false
	}

	// if the volume has attachments and the mask indicates that
	// only unattached volumes should be returned then omit this volume
	if len(vol.Attachments) > 0 &&
		!attachments.Attached() &&
		attachments.Unattached() {
		ctx.WithFields(lf).Debug("omitting attached volume")
		return false
	}

	return true
```

Or am I still missing something here?